File: services/jit-transcode-go/internal/transcoding/mp3_helpers.go
package transcoding

import (
	"crypto/md5"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/editframe/telecine/jit-transcode-go/internal/storage"
)

// IsMP3Url detects if a URL points to an MP3 file
func IsMP3Url(url string) bool {
	return strings.HasSuffix(strings.ToLower(url), ".mp3")
}

// GenerateMp3ConversionCacheKey creates a cache key for MP3 to MP4 conversion
func GenerateMp3ConversionCacheKey(originalMp3Url string) string {
	hash := md5.Sum([]byte(originalMp3Url))
	return fmt.Sprintf("mp3-conversions/%x.mp4", hash)
}

// ConvertMp3ToMp4AndCache converts an MP3 file to MP4 and caches the result
func ConvertMp3ToMp4AndCache(mp3Url string, storageProvider storage.StorageProvider) (string, error) {
	cacheKey := GenerateMp3ConversionCacheKey(mp3Url)

	exists, err := storageProvider.PathExists(cacheKey)
	if err != nil {
		return "", fmt.Errorf("failed to check cache existence: %w", err)
	}

	if exists {
		fmt.Printf("MP3 conversion cache hit: %s\n", mp3Url)
		return cacheKey, nil
	}

	fmt.Printf("Converting MP3 to MP4: %s\n", mp3Url)

	mp4Buffer, err := convertMp3ToMp4(mp3Url)
	if err != nil {
		return "", fmt.Errorf("failed to convert MP3: %w", err)
	}

	if err := storageProvider.WriteFile(cacheKey, mp4Buffer, storage.WriteOptions{
		ContentType: "video/mp4",
	}); err != nil {
		return "", fmt.Errorf("failed to cache MP3 conversion: %w", err)
	}

	fmt.Printf("Cached MP3 conversion: %s\n", cacheKey)
	return cacheKey, nil
}

// convertMp3ToMp4 performs the actual MP3 to MP4 conversion using FFmpeg
func convertMp3ToMp4(mp3Url string) ([]byte, error) {
	tmpDir := filepath.Join(os.TempDir(), "mp3-conversion")
	if err := os.MkdirAll(tmpDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create temp dir: %w", err)
	}

	hash := md5.Sum([]byte(mp3Url))
	inputPath := filepath.Join(tmpDir, fmt.Sprintf("%x-input.mp3", hash))
	outputPath := filepath.Join(tmpDir, fmt.Sprintf("%x-output.mp4", hash))

	defer os.Remove(inputPath)
	defer os.Remove(outputPath)

	if err := downloadFile(mp3Url, inputPath); err != nil {
		return nil, fmt.Errorf("failed to download MP3: %w", err)
	}

	cmd := exec.Command("ffmpeg",
		"-i", inputPath,
		"-c:a", "aac",
		"-b:a", "128k",
		"-ac", "2",
		"-ar", "48000",
		"-movflags", "+faststart",
		"-f", "mp4",
		"-y",
		outputPath,
	)

	if output, err := cmd.CombinedOutput(); err != nil {
		return nil, fmt.Errorf("ffmpeg conversion failed: %w, output: %s", err, string(output))
	}

	mp4Buffer, err := os.ReadFile(outputPath)
	if err != nil {
		return nil, fmt.Errorf("failed to read output file: %w", err)
	}

	return mp4Buffer, nil
}

// ResolveEffectiveTranscodingUrl resolves the effective URL for transcoding
// For MP3 URLs: downloads cached MP4 to temp file and returns file path
// For other URLs: returns original URL unchanged
func ResolveEffectiveTranscodingUrl(originalUrl string, storageProvider storage.StorageProvider) (string, error) {
	if !IsMP3Url(originalUrl) {
		return originalUrl, nil
	}

	cacheKey := GenerateMp3ConversionCacheKey(originalUrl)

	exists, err := storageProvider.PathExists(cacheKey)
	if err != nil {
		return "", fmt.Errorf("failed to check cache: %w", err)
	}

	if !exists {
		return "", fmt.Errorf("MP3 conversion not found. Call manifest endpoint first: %s", originalUrl)
	}

	tmpDir := filepath.Join(os.TempDir(), "mp3-transcoding")
	if err := os.MkdirAll(tmpDir, 0755); err != nil {
		return "", fmt.Errorf("failed to create temp dir: %w", err)
	}

	hash := md5.Sum([]byte(originalUrl))
	tempFilePath := filepath.Join(tmpDir, fmt.Sprintf("%x.mp4", hash))

	info, err := os.Stat(tempFilePath)
	if err == nil {
		ageMinutes := time.Since(info.ModTime()).Minutes()
		if ageMinutes < 60 {
			fmt.Printf("Reusing cached MP4 temp file: %s\n", tempFilePath)
			return tempFilePath, nil
		}
	}

	fmt.Printf("Downloading cached MP4 for transcoding: %s -> %s\n", cacheKey, tempFilePath)

	reader, err := storageProvider.CreateReadStream(cacheKey)
	if err != nil {
		return "", fmt.Errorf("failed to read cached MP4: %w", err)
	}
	defer reader.Close()

	outFile, err := os.Create(tempFilePath)
	if err != nil {
		return "", fmt.Errorf("failed to create temp file: %w", err)
	}
	defer outFile.Close()

	if _, err := io.Copy(outFile, reader); err != nil {
		return "", fmt.Errorf("failed to copy cached MP4: %w", err)
	}

	return tempFilePath, nil
}

// ValidateMp3Rendition validates that the rendition is supported for MP3 files
func ValidateMp3Rendition(rendition string) bool {
	return rendition == "audio"
}

// downloadFile downloads a file from a URL
func downloadFile(url, filepath string) error {
	cmd := exec.Command("curl", "-L", "-o", filepath, url)
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to download file: %w", err)
	}
	return nil
}

File: telecine/services/jit-transcoding/mp3-helpers.ts
import * as crypto from "node:crypto";
import { storageProvider } from "@/util/storageProvider.server";
import { getFileDuration } from "@/transcode/src/jit/transcoding-service";
import { convertMp3ToMp4 } from "@/transcode/src/jit/audio-transcoder";
import * as fs from "node:fs/promises";
import * as path from "node:path";
import * as os from "node:os";
import { createWriteStream } from "node:fs";
import { calculateSegmentDurations } from "@/transcode/src/jit/calculateSegmentDurations";

// Helper to generate cache key for MP3 → MP4 conversion
export function generateMp3ConversionCacheKey(originalMp3Url: string): string {
  const urlHash = crypto.createHash("md5").update(originalMp3Url).digest("hex");
  return `mp3-conversions/${urlHash}.mp4`;
}

// Helper to convert MP3 to MP4 and cache the result
export async function convertMp3ToMp4AndCache(mp3Url: string): Promise<string> {
  const cacheKey = generateMp3ConversionCacheKey(mp3Url);

  // Check if conversion already exists in storage provider
  const cacheExists = await storageProvider.pathExists(cacheKey);

  if (cacheExists) {
    console.log(`MP3 conversion cache hit: ${mp3Url}`);
    return cacheKey;
  }

  console.log(`Converting MP3 to MP4: ${mp3Url}`);

  try {
    // Convert MP3 to MP4
    const mp4Buffer = await convertMp3ToMp4(mp3Url);

    // Store in storage provider instead of local disk
    await storageProvider.writeFile(cacheKey, mp4Buffer, {
      contentType: "video/mp4",
    });

    console.log(`Cached MP3 conversion: ${cacheKey}`);
    return cacheKey;
  } catch (error) {
    console.error(`Failed to convert MP3 to MP4: ${mp3Url}`, error);
    throw error;
  }
}

// Helper to resolve the effective URL for transcoding
// For MP3 URLs: downloads cached MP4 to temp file and returns file path
// For other URLs: returns original URL unchanged
export async function resolveEffectiveTranscodingUrl(
  originalUrl: string,
): Promise<string> {
  if (!isMP3Url(originalUrl)) {
    return originalUrl;
  }

  const cacheKey = generateMp3ConversionCacheKey(originalUrl);

  // Check if conversion exists in storage provider
  const cacheExists = await storageProvider.pathExists(cacheKey);

  if (!cacheExists) {
    throw new Error(
      `MP3 conversion not found. Call manifest endpoint first: ${originalUrl}`,
    );
  }

  // Download cached MP4 to temporary file for transcoding
  const tempDir = path.join(os.tmpdir(), "mp3-transcoding");
  await fs.mkdir(tempDir, { recursive: true });

  const urlHash = crypto.createHash("md5").update(originalUrl).digest("hex");
  const tempFilePath = path.join(tempDir, `${urlHash}.mp4`);

  // Check if temp file already exists and is recent (avoid re-downloading)
  try {
    const stats = await fs.stat(tempFilePath);
    const ageMinutes = (Date.now() - stats.mtime.getTime()) / (1000 * 60);
    if (ageMinutes < 60) {
      // Reuse temp file if less than 1 hour old
      console.log(`Reusing cached MP4 temp file: ${tempFilePath}`);
      return tempFilePath;
    }
  } catch {
    // File doesn't exist, continue with download
  }

  console.log(
    `Downloading cached MP4 for transcoding: ${cacheKey} -> ${tempFilePath}`,
  );

  try {
    // Download from storage provider to temp file
    const cachedStream = await storageProvider.createReadStream(cacheKey);
    const writeStream = createWriteStream(tempFilePath);

    await new Promise<void>((resolve, reject) => {
      cachedStream.pipe(writeStream);
      writeStream.on("finish", resolve);
      writeStream.on("error", reject);
      cachedStream.on("error", reject);
    });

    return tempFilePath;
  } catch (error) {
    console.error(`Failed to download cached MP4: ${cacheKey}`, error);
    throw new Error(
      `Failed to access cached MP4 for transcoding: ${originalUrl}`,
    );
  }
}

// Helper to generate MP3 manifest
export async function generateMp3Manifest(mp3Url: string, baseUrl: string) {
  // Ensure MP3 is converted to MP4 and cached
  await convertMp3ToMp4AndCache(mp3Url);

  // Get a local file path for duration extraction (downloads from storage provider to temp file)
  const localMp4Path = await resolveEffectiveTranscodingUrl(mp3Url);

  // Get duration from the local MP4 file using the correct function for local files
  const duration = await getFileDuration(localMp4Path);
  const durationMs = duration * 1000;

  // Calculate AAC frame-aligned segment durations (returns milliseconds)
  const targetSegmentDurationMs = 15000; // 15 seconds target
  const audioSegmentDurations = calculateSegmentDurations(
    durationMs,
    targetSegmentDurationMs,
    {
      mediaType: "audio",
    },
  );

  return {
    version: "1.0",
    type: "jit-audio",
    sourceUrl: mp3Url, // Always use original MP3 URL
    duration: duration,
    durationMs: durationMs,
    baseUrl: baseUrl,

    // Audio-only renditions for MP3
    videoRenditions: [], // No video for MP3
    audioRenditions: [
      {
        id: "audio",
        channels: 2,
        sampleRate: 48000,
        bitrate: 128000,
        codec: "mp4a.40.2", // AAC-LC
        container: "audio/mp4",
        mimeType: 'audio/mp4; codecs="mp4a.40.2"',
        segmentDuration: 15, // 15 seconds target
        segmentDurationMs: targetSegmentDurationMs,
        segmentDurationsMs: audioSegmentDurations,
        language: "en",
      },
    ],

    // Segment URL templates use ORIGINAL MP3 URL - no internal paths exposed
    endpoints: {
      initSegment: `${baseUrl}/api/v1/transcode/{rendition}/init.m4s?url=${encodeURIComponent(mp3Url)}`,
      mediaSegment: `${baseUrl}/api/v1/transcode/{rendition}/{segmentId}.m4s?url=${encodeURIComponent(mp3Url)}`,
    },

    jitInfo: {
      parallelTranscodingSupported: true,
      expectedTranscodeLatency: 100, // Faster for pre-converted MP3s
      segmentCount: audioSegmentDurations.length,
    },
  };
}

// Helper to detect if a URL is an MP3 file
export function isMP3Url(url: string): boolean {
  return url.toLowerCase().endsWith(".mp3");
}

// Helper to validate MP3 rendition
export function validateMp3Rendition(rendition: string): boolean {
  return rendition === "audio";
}

// Helper to validate 15s time alignment for MP3
export function validateMp3TimeAlignment(startTimeMs: number): {
  isValid: boolean;
  nearestValidTime?: number;
} {
  const isValid = startTimeMs % 15000 === 0;
  const nearestValidTime = Math.round(startTimeMs / 15000) * 15000;

  return {
    isValid,
    nearestValidTime: isValid ? undefined : nearestValidTime,
  };
}
